### PR TITLE
feat(@toss/react): add Portal Component

### PR DESCRIPTION
**Pull request by ssi02014**
## Overview

I worked on a Portal component to utilize the `createPortal` that React provides.

Please review and provide feedback on whether this component is useful and appropriate for "toss/slash". 🙏

For reference, I've added `@types/react-dom` to `@toss/react`.

```tsx
const Example = () => {
  return (
    <Portal>
      <Modal />
    </Portal>
  );
};
```

## PR Checklist

- [x] I read and included theses actions below

1. I have read the [Contributing Guide](https://github.com/toss/slash/blob/main/.github/CONTRIBUTING.md)
2. I have written documents and tests, if needed.

### CODE REVIEW

File: packages/react/react/src/components/Portal/index.spec.tsx
Comment (ChanhyukPark-Tech) [inline packages/react/react/src/components/Portal/index.spec.tsx:31]: I think many users who use `Portal` component do now want this feature😄
How about writing this part in the `document` because I think it's a side effect

File: packages/react/react/src/components/Portal/Portal.tsx
Comment (ChanhyukPark-Tech) [inline packages/react/react/src/components/Portal/Portal.tsx:18]:
```suggestion
  const isMounted = useIsMounted();

  /**
   * With this code, it is possible to solve the "window is not defined" and "Hydration Error" that can occur in SSR.
   */
  if (!isMounted) {
    return <></>;
  }
```

I think this is more appropriate😄

### Usage

image

Comment (ssi02014) [reply packages/react/react/src/components/Portal/Portal.tsx:18]: That's something I missed..!!!

Thank you for the feedback. I agree with your opinion 👍

I'll fix that code in a moment🙆‍♂️

File: packages/react/react/src/components/Portal/index.spec.tsx
Comment (ssi02014) [reply packages/react/react/src/components/Portal/index.spec.tsx:31]: I understood your comment as adding what you mentioned in the ".md" document. 

Am I understanding your comment correctly?!

If this is correct, I will fix this as well!

Thank you for the feedback 🙂🙂

Comment (ChanhyukPark-Tech) [reply packages/react/react/src/components/Portal/index.spec.tsx:31]: yes I think it is good to be in .md file 😄

**Comment by ssi02014** [comment]
@ChanhyukPark-Tech 
Fixed it! 🙏

File: packages/react/react/src/components/Portal/Portal.tsx
Comment (ChanhyukPark-Tech) [inline packages/react/react/src/components/Portal/Portal.tsx:6]: How about trying to find the parent's node to render using the `ref` value instead of the `id`(string)?

Looking at the various implemented libraries, it seems to be implemented as `ref`, not as the string value id.

If it is implemented as an `id`, which is a normal magic string, it is difficult to find which file has the element with the `that id`, and it is likely to cause side effects.

However, if you manage it with `ref`, the location of the parent node can be directly identified on the code, so I think the developer experience will be good 

ref

[chakra-ui](https://chakra-ui.com/docs/components/portal/props)
[mui](https://mui.com/base-ui/react-portal/components-api/)

Comment (ssi02014) [reply packages/react/react/src/components/Portal/Portal.tsx:6]: Thanks for the great feedback. 😀

Definitely, managing by `id(string)`  is more likely to cause side effects, and it's a bad developer experience. 

I will think about a Portal Component managed by `ref`. 

Additionally, the default behavior of "mui" is to add a span tag to the "body" tag, and "chakra ui" is to add a div tag with "chakra-portal" as class to the body tag.

The default behavior of chakra ui seems more appropriate to me, and I will try to design similarly to it.

Comment (ssi02014) [reply packages/react/react/src/components/Portal/Portal.tsx:6]: @ChanhyukPark-Tech 
Oh sorry, I was mistaken about the MUI default behavior.

**Comment by ssi02014** [comment]
@ChanhyukPark-Tech 
I agree with you and have modified the overall structure and functionality of the "Portal" component. 😀

The Portal component now renders the Portal Node in `document.body` by default.

However, if the user passes in the `containerRef` prop, it is possible to render the Portal Node in a DON Node other than document.body. 

Additionally, it supports nested portals. (But, I don't think nesting is the primary purpose of this component.)

Thank you again for the great feedback 🙏

File: packages/react/react/src/components/Portal/Portal.en.md
Comment (ChanhyukPark-Tech) [inline packages/react/react/src/components/Portal/Portal.en.md:20]:
```suggestion
  children: React.ReactNode;
```

Comment (ssi02014) [reply packages/react/react/src/components/Portal/Portal.en.md:20]: @ChanhyukPark-Tech 
OMG.. Thank you. I'll fix it right now

File: packages/react/react/src/components/Portal/Portal.tsx
Comment (ChanhyukPark-Tech) [inline packages/react/react/src/components/Portal/Portal.tsx:23]: What do you think is better?
If you're going to set it as the default anyway, how about setting it from the beginning?

```suggestion
function RenderPortal({ children, className = PORTAL_DEFAULT_CLASS, containerRef }: PortalProps) {
  const { parentPortal } = useContext(portalContext);

  const getPortalNode = useCallback(
    (mountNode: HTMLElement) => {
      const portalNode = mountNode.ownerDocument.createElement('div');
      portalNode.classList.add(className);
```

File: packages/react/react/src/components/Portal/Portal.en.md
Comment (ChanhyukPark-Tech) [reply packages/react/react/src/components/Portal/Portal.en.md]: You can easily fix it just by pressing the `commit suggestion` button on my comment!

Comment (ssi02014) [reply packages/react/react/src/components/Portal/Portal.en.md]: @ChanhyukPark-Tech 
.....😂😂😂😂😂😂😂😂😂
Thank you so much for these tip. I was being an idiot and modifying the code myself every time, registering commits one by one.

File: packages/react/react/src/components/Portal/Portal.tsx
Comment (ssi02014) [reply packages/react/react/src/components/Portal/Portal.tsx:76]: @ChanhyukPark-Tech 
My intention is simply that Portal's return type is inferred to be 'JSX.Element | null'. 

If you use Fragment, it will be inferred as 'JSX.Element' instead of 'JSX.Element | null'.

But neither is a big deal functionally... 👍

**Review by ChanhyukPark-Tech — APPROVED** [review]
thanks @ssi02014 

I think this component is good.
other slash maintainer, please double check.

File: packages/react/react/src/components/Portal/Portal.tsx
Comment (manudeli) [reply packages/react/react/src/components/Portal/Portal.tsx:66]: neither is a big deal functionally... but Isn't it better to express it clearly to user: `this component can return nothing` by using null?

https://react.dev/learn/conditional-rendering#conditionally-returning-nothing-with-null

I could understand your intention, but I want to know why It's better `JSX.Element` than `JSX.Element | null`.
because I thought it's more clear,`JSX.Element | null`. I want to know case that why you need just `JSX.Element`. (I'm not sure if this is correct, so I'm curious about what many people think. 
@ssi02014, @ChanhyukPark-Tech, @raon0211, @HoseungJang etc.)

Comment (ssi02014) [reply packages/react/react/src/components/Portal/Portal.tsx:66]: @manudeli
Thanks for the feedback. 🙏

I too was initially torn between null and Fragment.

My initial intent was simply that the Portal would render based on whether it was mounted or not, so I assumed that it would always render unless there were special circumstances, such as an error.

So, I felt it was awkward to have null as a return type, and since neither null nor Fragment were functionally problematic, I chose Fragment.

However, I'm going to defer to the majority here because I don't think this debate is a functional issue, but a matter of preference. 🙇‍♂️

**Pull request closed without merging**
